web/src/message_events.ts

```diff
@@ -287,14 +295,37 @@ export function rewire_update_views_filtered_on_message_property(
     update_views_filtered_on_message_property = value;
 }
 
-export function insert_new_messages(
-    raw_messages: RawMessage[],
-    sent_by_this_client: boolean,
-    deliver_locally: boolean,
-): Message[] {
-    const messages = raw_messages.map((raw_message) =>
-        message_helper.process_new_message(raw_message, deliver_locally),
-    );
+export type InsertNewMessagesOpts = {
+    sent_by_this_client: boolean;
+} & (
+    | {
+          type: "server_message";
+          raw_messages: RawMessage[];
+      }
+    | {
+          type: "local_message";
+          raw_messages: LocalMessage[];
+      }
+);
+
+export function insert_new_messages(opts: InsertNewMessagesOpts): Message[] {
+    const deliver_locally = opts.type === "local_message";
+    let messages: Message[] = [];
+    if (opts.type === "server_message") {
+        messages = opts.raw_messages.map((raw_message) =>
+            message_helper.process_new_message({
+                type: opts.type,
+                raw_message,
+            }),
+        );
+    } else {
+        messages = opts.raw_messages.map((raw_message) =>
+            message_helper.process_new_message({
+                type: opts.type,
+                raw_message,
+            }),
+        );
```

Comment:
The if/else clauses seem identical here?

Reply:
Ahh, but it stops being one commit later. OK.

Reply:
I also had to do this in this commit because I'd get this kind of typing error otherwise:

```
 Types of property 'type' are incompatible.
      Type '"server_message" | "local_message"' is not assignable to type '"local_message"'.
        Type '"server_message"' is not assignable to type '"local_message"'.
```